### ISSUE TITLE
Photo damage classifier v2: ~87% FP rate at production threshold 0.20

### ISSUE THREAD

**Issue by nikit34** [nikit34/olx-car-parser#1]
## Summary

Audit on a random 100-listing sample of the **flagged** segment (`photo_damage_p ≥ 0.20`) shows that **~87%** of the flags are false positives — pristine dealer photos, interior/detail shots, and high-contrast reflections. The current production threshold (`0.20`, `src/parser/photo_damage.py:39`) was tuned on a 51-listing gold benchmark with a strong recall objective (F1=0.818, R=100%), but on the actual production distribution (6 492 listings as of 2026-05-01) the precision collapses.

`flagged P ≥ 0.20` covers **3 250 listings** (50.1% of all photo-verified). At the audit's effective precision ≈ **13%** that means roughly **~420 true positives and ~2 800 false positives** sitting in the flagged bucket. Alerts currently hide most of this through `_blocking_deal_reason` cross-checks against text `damage_severity`, but the standalone photo signal is unreliable.

## Methodology

1. Pulled a uniform random sample of 100 active flagged listings from the host DB:
   ```sql
   SELECT olx_id, url, title, json_extract(llm_extras, '\$.photo_damage_p') AS max_p
   FROM listings
   WHERE is_active = 1
     AND CAST(json_extract(llm_extras, '\$.photo_damage_p') AS REAL) >= 0.20
     AND (url LIKE '%standvirtual%' OR url LIKE '%olx.pt%')
   ORDER BY RANDOM() LIMIT 100;
   ```
2. For each listing: fetched all photos via `src.parser.photo_fetch.fetch_photos`, ran `DamageClassifier` on every photo, kept the **single highest-scoring photo** per listing — i.e. the one that drove the listing's flag.
3. Manually labelled the 23 photos sampled across the score distribution (top 10 + every 5–10th down to the threshold) as TP / borderline TP / FP.

(End-to-end run took 6 min on the scrape host's MPS device.)

## Results

| rank | p_damaged | listing | verdict |
|---|---|---|---|
| 001 | 0.999 | BMW M4 GTS rear (carbon spoiler, dealer shot) | **FP** |
| 002 | 0.998 | Cupra Born front (dark blue, harsh sun) | **FP** |
| 003 | 0.997 | Tesla Model 3 rear-side (clean) | **FP** |
| 004 | 0.997 | Renault Laguna front-quarter (faded clear-coat) | borderline TP |
| 005 | 0.994 | Volvo V40 (clean) | **FP** |
| 006 | 0.989 | Mercedes E-Class wagon (wet pavement reflections) | **FP** |
| 007 | 0.985 | trunk shot with crumpled foil bag | **FP** |
| 008 | 0.982 | Mitsubishi Outlander (cobblestone shadows) | **FP** |
| 009 | 0.978 | Peugeot 3008 dealer (washing hose on ground) | **FP** |
| 010 | 0.970 | Tesla wheel close-up | **FP** |
| 015 | 0.925 | Tesla Model 3 white (clean) | **FP** |
| 020 | 0.895 | Toyota C-HR red (foliage reflections) | **FP** |
| 030 | 0.825 | Mercedes C-Class indoor (clean) | **FP** |
| 040 | 0.710 | Opel Corsa B (oxidised paint, faded headlight) | borderline TP |
| **050** | **0.564** | **Nissan engine bay, front bumper REMOVED** | **TP** |
| 060 | 0.461 | Mini engine bay | **FP** |
| 070 | 0.408 | Fiat 500 wheel close-up | **FP** |
| 080 | 0.346 | rear AC vents (interior) | **FP** |
| 085 | 0.327 | trunk loaded with bags | **FP** |
| 090 | 0.288 | sunroof from inside | **FP** |
| 095 | 0.242 | Volvo V40 grey rear (clean parking shot) | **FP** |
| 098 | 0.200 | Audi A1 blue rear (clean) | **FP** |
| 100 | 0.002 | Toyota Auris — *cron flagged ≥0.20, rerun disagrees* | FP (re-run drift) |

**Tally: 1 clear TP, 2 borderline TPs, 20 clear FPs across the score range.**

The single clear TP (Nissan with bumper off) scored only **0.564** — weaker signal than half the FPs above it. SNR is inverted: harsh sunlight on a black BMW M4 reads as more damaged than an actual stripped front end.

## Root causes (hypothesis)

1. **Pristine dealer photos with reflections / harsh sun on dark glossy panels** dominate the high-confidence FP bucket. Carbon trim on the BMW M4, sun glare on the Cupra Born — the model reads high-contrast patches as paint mismatch.
2. **Interior / engine-bay / wheel close-ups are out-of-distribution** for a binary "damaged car / clean car" classifier trained on full-vehicle exterior shots. They should produce ≈0.5 ("not sure"); instead they confidently produce 0.3–0.99. Suggests the training set under-represented these viewpoints.
3. **Wet pavement, cobblestones, foliage shadows** all look like patchy paint at low resolution.
4. **Listing-level aggregation by `max(p_damaged)`** amplifies any single hot photo. With 50 photos per listing on StandVirtual, one weirdly-lit shot is enough to flag the whole car.

## Recommendations

In rough order of impact / cost:

### Short term (no retrain)

1. **Raise threshold to ~0.50** OR **require ≥2 photos with p ≥ 0.30** before flagging the listing. Cuts the bottom three quarters of the current flagged set, where the audit found zero TPs.
2. **Pre-filter by photo type before classifying.** A cheap CLIP zero-shot pass ("photo of car exterior" vs "photo of car interior / engine bay / wheel close-up") to skip OOD viewpoints. The damage_classifier_v2 was clearly never meant to score interiors.
3. **Persist all per-photo `p_damaged` values in `llm_extras`**, not just the listing-level `max`. Without per-photo scores we can't even retrain on production feedback. Cheap schema change; payoff is large.

### Medium term

4. **Re-train v3** on a balanced mix:
   - DrBimmer + the existing OLX hand-labels (TP source)
   - **Production false-positives** sampled from this audit's flagged-but-clean set (FP source — the model's actual blind spots)
   - Stratified by photo viewpoint (exterior / interior / engine bay / wheel) and by lighting condition (sun glare / wet / shadow / studio)
5. **Per-viewpoint head** instead of one binary classifier: separate damage heads for exterior/interior/mechanical, only the exterior head feeds the listing-level signal.

### Why this isn't urgent (yet)

Alerts compensate for the noise via `_blocking_deal_reason` requiring text `damage_severity` agreement, so the user-facing FP rate is far lower than 87%. But:
- The `photo_damage_p` field is intended to be a useful standalone signal (e.g. for analytics, sub-segment damage rates).
- Eventually we'd want to lower the alert system's reliance on text signals (LLM extraction misses obvious salvage cars too); that requires a precise photo signal we don't currently have.

## Reproduction

The audit script and helpers are reusable:

```bash
# On the dev machine, query the production DB on the scrape host:
sqlite3 -separator $'\t' /Users/anastasia/olx-car-parser/data/olx_cars.db \
  \"SELECT olx_id, url, replace(title, char(9), ' '), json_extract(llm_extras, '\$.photo_damage_p') AS max_p
   FROM listings WHERE is_active = 1
     AND CAST(json_extract(llm_extras, '\$.photo_damage_p') AS REAL) >= 0.20
     AND (url LIKE '%standvirtual%' OR url LIKE '%olx.pt%')
   ORDER BY RANDOM() LIMIT 100\" > listings.tsv

# Run on the scrape host (has torch + weights + photo cache):
.venv/bin/python audit_flagged.py   # (script body in audit-2026-05-01 transcript)
# Outputs /tmp/audit/manifest.csv + /tmp/audit/view/{rank}_p{score}_{olx_id}.jpg
```

## Related

- `src/parser/photo_damage.py:39` — `DEFAULT_THRESHOLD = 0.20`
- `src/cli.py` `verify-photos` command — the cron consumer
- v2 weights file `damage_classifier_v2.pt` — 94 MB, trained on DrBimmer + OLX hand-labels

**Comment by nikit34** [comment]
Recommendations split into actionable child issues:

**Short term (no retrain, ship-ready):**
- #2 — raise threshold to 0.50 or require multi-photo agreement (S, ½ day)
- #3 — CLIP pre-filter for non-exterior photos (M, ½–1 day)
- #4 — persist per-photo `p_damaged` in `llm_extras` (S, 2–4h)

**Medium term (retrain):**
- #5 — train v3 with mined production false-positives (L, 3–5 days; depends on #3 + #4)

**Long term (architecture):**
- #6 — v4 per-viewpoint heads (XL, 1–2 weeks; depends on #5; gated on v3 precision flatlining <0.65)

Suggested order: #4 → #2 → #3 → #5 → reassess for #6.

This issue stays open as the tracking parent.

**Comment by nikit34** [comment]
## Production validation — #2 / #3 / #4 measurably reduce flagged FP rate

After the workflow's verify-photos step processed 717 fresh listings under the new code, the back-of-envelope numbers from comparing the two flagging rules in-place:

| rule | flagged count | % flagged |
|---|---|---|
| v2 max-rule (`photo_damage_p ≥ 0.20`) | 235 / 717 | **32.8 %** |
| new multi-photo agreement (≥2 photos at ≥0.30) | 69 / 717 | **9.6 %** |

**−70.6 % flagged-listing reduction** (235 → 69). 166 listings that v2 max-rule would flag are *not* flagged under the new rule. Zero listings added to flagged that v2 didn't (sanity — new rule is strictly more conservative, recall risk is one-sided).

CLIP exterior pre-filter (#3) is doing real work too: of 717, 97.2 % had at least one photo filtered out as OOD; average drop was 13 photos → 5.9 exterior (≈55 % cut). Three listings ended up with `n_exterior=0` (all photos were interiors / engine bays / wheel close-ups — classifier never ran on them, no flag).

Headline: at the audit's ~13 % precision baseline, cutting 70.6 % of flags pushes effective precision-at-flag toward ~44 % under the new rule, *if* the cut preserves audit-level TPs (Nissan-without-bumper signals were on multi-photo listings, so plausibly yes). A fresh ground-truth audit on the 69 newly-flagged listings would confirm.

### Took three dispatches to validate

The first two production runs of #3 looked green but wrote zero new fields — both are lessons for the workflow's masking behaviour:

1. `pyproject.toml` didn't list `transformers` (only `requirements.txt` did), so CI's `pip install -e .` skipped it → `ModuleNotFoundError` at verify-photos startup, masked by `continue-on-error: true` on the step. Fixed in `b13e6c8`.
2. `ThreadPoolExecutor(workers=4)` ran concurrent forward passes through the single MPS-resident CLIP + ResNet50 → `Trace/BPT trap: 5` (SIGTRAP, exit 133), also masked by `continue-on-error`. Fixed in `2e649c8` (per-instance `threading.Lock` around inference).

Worth filing as a separate issue: `continue-on-error: true` on the verify-photos step swallows errors that produce zero output — a step that runs for 30 s and writes nothing should at least warn. Adding a "saved=0 listings → ::warning::" guard would have caught both bugs in the first run.

### Where the wins land

Alerts and dashboard still threshold on the old `photo_damage_p` (max), so end-users haven't seen the precision improvement yet — these are the consumers that need to flip to `photo_damage_flagged` (new field) to inherit the reduction. Filing a follow-up issue for that migration.
